compiler/rustc_session/src/utils.rs

```diff
@@ -91,3 +100,67 @@ impl CanonicalizedPath {
         &self.original
     }
 }
+
+/// A path that should be invalidated when the file that it points to has changed.
+#[derive(Clone, Debug)]
+pub struct HashedFilePath {
+    path: PathBuf,
+}
+
+impl HashedFilePath {
+    pub fn new(path: PathBuf) -> Self {
+        Self { path }
+    }
+
+    pub fn as_path(&self) -> &Path {
+        self.path.as_path()
+    }
+}
+
+impl From<HashedFilePath> for PathBuf {
+    fn from(path: HashedFilePath) -> Self {
+        path.path
+    }
+}
+
+impl DepTrackingHash for HashedFilePath {
+    fn hash(
+        &self,
+        hasher: &mut DefaultHasher,
+        _error_format: ErrorOutputType,
+        _for_crate_hash: bool,
+    ) {
+        // Q1: Should we also hash the filepath itself?
+        // Q2: Maybe we could hash directly into DefaultHasher?
+        let hash = hash_file(&self.path);
+        match hash {
+            Ok(hash) => std::hash::Hash::hash(&hash[..], hasher),
+            Err(_) => {
+                // The file does not exist or could not be hashed. Use a special marker value for
+                // the hash.
+                std::hash::Hash::hash(
+                    &[0; <CoreWrapper<Md5Core> as OutputSizeUser>::OutputSize::USIZE],
+                    hasher,
+                );
+            }
+        };
+    }
+}
+
+fn hash_file(path: &Path) -> std::io::Result<Output<Md5>> {
+    let mut hasher = Md5::default();
```

Comment:
Hmm, AFAIK `StableHasher` is optimized for frequent and small updates, while here we want to hash large chunks of bytes.

I originally wanted to use Blake3, but decided on MD5 to avoid adding a new dependency. I created a small benchmark, which hashed a 4GiB profile file with random data (primed in file cache, so hopefully without much I/O overhead) with 128 KiB block size:
```
MD5: 9200 - 10000ms
BLAKE3: 1099 - 1100ms
StableHasher: 1687 - 1744ms
```

BLAKE3 looks like a good choice, especially since it was designed for use-cases like hashing files. But the stable hasher also isn't half bad.

Reply:
(Btw I also tried swapping the hashing algorithm for source files for blake3, but it didn't really help. Well, most source files don't have 4 GiB :) ).

Reply:
XXH3 would be the ideal choice ;)

Do we already have a dependency on BLAKE3? I agree that it looks like a great choice!

Reply:
We don't, I would have to add it to `rustc_session`. It uses `CC0 1.0`/`Apache 2.0` license.

Reply:
Btw, when we do PGO for LLVM in CI, all the files have 16 GiB, but after they are combined into a single file, it's just under 30 MiB. For `rustc` it's 80 MiB. So maybe large optimizations here are unnecessary and we can just use `StableHasher` to avoid a new dependency.

Reply:
> all the files have 16 GiB
😅 

I'd just go with StableHasher for now, I think. We can always switch later.